### PR TITLE
manage stop time headsign

### PR DESCRIPTION
**Pull request by datanel**
bench on ntfs2ntfs fr-idf

| | without headsign | with headsign
| --- | --- | ---
| Mem | 1,550GB | 1,563GB
| Time | 35s | 33s

### CODE REVIEW

File: src/model.rs [hunk 1]
```diff
@@ -48,6 +48,8 @@ pub struct Collections {
     pub trip_properties: CollectionWithId<TripProperty>,
     pub geometries: CollectionWithId<Geometry>,
     pub admin_stations: Collection<AdminStation>,
+    #[serde(skip)]
+    pub stop_time_headsigns: HashMap<(Idx<VehicleJourney>, u32), String>,
```
[the remark this answers is not in the document]
Comment (datanel) [reply src/model.rs]: beacause 
```
error[E0277]: the trait bound `collection::Idx<objects::VehicleJourney>: serde::Serialize` is not satisfied
  --> src/model.rs:51:5
   |
51 |     pub stop_time_headsigns: HashMap<(Idx<VehicleJourney>, u32), String>,
   |     ^^^ the trait `serde::Serialize` is not implemented for `collection::Idx<objects::VehicleJourney>`
   |
   = note: required because of the requirements on the impl of `serde::Serialize` for `(collection::Idx<objects::VehicleJourney>, u32)`
   = note: required because of the requirements on the impl of `serde::Serialize` for `std::collections::HashMap<(collection::Idx<objects::VehicleJourney>, u32), std::string::String>`
   = note: required by `serde::ser::SerializeStruct::serialize_field`
```

and why do I need to serialize it?

[the remark this answers is not in the document]
Comment (TeXitoi) [reply src/model.rs]: OK.

File: src/model.rs [hunk 2]
```diff
@@ -74,17 +76,55 @@ impl Collections {
...
+        }
+
+        let sp_idx_to_id = stop_points.idx_to_id();
+        let vj_idx_to_id = vehicle_journeys.idx_to_id();
+
         self.stop_points.merge(stop_points)?;
+        self.vehicle_journeys.merge(vehicle_journeys)?;
+
+        // Update stop point idx in stop times
+        let mut vjs = self.vehicle_journeys.take();
```
Comment (TeXitoi) [inline src/model.rs]: That's done on the merged view, it will not work: old vjs will have stop_points on the new stop points.

We need a test that merge 2 models and check that the vj are as expected.

File: src/model.rs [hunk 3]
```diff
@@ -74,17 +76,62 @@ impl Collections {
...
+                    st.stop_point_idx = new_idx;
+                }
+            }
+        }
+        vehicle_journeys = CollectionWithId::new(vjs)?;
+        self.vehicle_journeys.merge(vehicle_journeys)?;
+
+        // Update vehicle journey idx
+        let mut new_stop_time_headsigns = HashMap::new();
+        for ((old_vj_idx, sequence), headsing) in &stop_time_headsigns {
```
Comment (pbougue) [inline src/model.rs]: typo: "headsi`gn`" not ":speaking_head: :notes: "

File: tests/merge_ntfs.rs [hunk 4]
```diff
@@ -47,6 +52,36 @@ fn merge_collections_ok() {
     assert_eq!(collections.lines.len(), 6);
     assert_eq!(collections.routes.len(), 8);
     assert_eq!(collections.vehicle_journeys.len(), 8);
+    assert_eq!(collections.stop_time_headsigns.len(), 1);
```
Comment (pbougue) [inline tests/merge_ntfs.rs]: Don't you want to check to content also?

File: src/ntfs/mod.rs [hunk 5]
```diff
@@ -534,6 +549,7 @@ mod tests {
 
             read::manage_stop_times(&mut collections, path).unwrap();
             assert_eq!(collections.vehicle_journeys, vehicle_journeys);
+            assert_eq!(headsigns, collections.stop_time_headsigns);
```
[the remark this answers is not in the document]
Comment (datanel) [reply src/ntfs/mod.rs]: because here we do a write and a read

File: src/model.rs [hunk 6]
```diff
@@ -74,17 +76,62 @@ impl Collections {
...
+                .collect()
+        }
+
+        let sp_idx_to_id = idx_to_id(&stop_points);
+        let vj_idx_to_id = idx_to_id(&vehicle_journeys);
+
         self.stop_points.merge(stop_points)?;
+
+        // Update stop point idx in new stop times
+        let mut vjs = vehicle_journeys.take();
```
Comment (datanel) [inline src/model.rs]: this is be done in another PR https://github.com/CanalTP/navitia_model/pull/146

File: tests/merge_ntfs.rs (hunk 4 above)
Comment (datanel) [reply tests/merge_ntfs.rs]: done below

File: src/ntfs/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (datanel) [reply src/ntfs/mod.rs]: in these tests we always do a "write" and a "read"
model -> write ntfs -> read ntfs -> model and we compare the input with the output
